chiron/mcmc.py

```diff
@@ -112,103 +170,802 @@ def run(
+        )
+
+    def _update_stepsize(self):
+        """
+        Update the displacement_sigma to reach a target acceptance probability of 0.5.
+        """
+        acceptance_ratio = self.n_accepted / self.n_proposed
+        if acceptance_ratio > 0.6:
+            self.displacement_sigma *= 1.1
+        elif acceptance_ratio < 0.4:
+            self.displacement_sigma /= 1.1
+
+    def _propose(
+        self,
+        current_sampler_state: SamplerState,
+        current_thermodynamic_state: ThermodynamicState,
+        current_reduced_pot: float,
+        current_nbr_list: Optional[PairsBase] = None,
+    ) -> Tuple[SamplerState, ThermodynamicState, float, float, Optional[PairsBase]]:
         """
-        Decide whether to accept or reject the move based on the acceptance probability.
+        Implement the logic specific to displacement changes.
 
         Parameters
         ----------
-        probability : float
-            Acceptance probability.
+        current_sampler_state : SamplerState, required
+            Current sampler state.
+        current_thermodynamic_state : ThermodynamicState, required
+            Current thermodynamic state.
+        current_reduced_pot : float, required
+            Current reduced potential.
+        current_nbr_list : Optional[PairsBase]
+            Neighbor list associated with the current state.
 
         Returns
         -------
-        bool
-            Boolean indicating if the move is accepted.
+        proposed_sampler_state : SamplerState
+            Proposed sampler state.
+        proposed_thermodynamic_state : ThermodynamicState
+            Proposed thermodynamic state.
+        proposed_reduced_pot : float
+            Proposed reduced potential.
+        log_proposal_ratio : float
+            Log proposal ratio.
+        proposed_nbr_list : PairsBase
+            Proposed neighbor list. If not defined, this will be None.
         """
-        import jax.numpy as jnp
 
-        return jnp.random.rand() < probability
+        # create a mask for the atom subset: if a value of the mask is 0
+        # the particle won't move; if 1 the particle will be moved
+        if self.atom_subset is not None and self.atom_subset_mask is None:
+            import jax.numpy as jnp
+
+            self.atom_subset_mask = jnp.zeros(current_sampler_state.n_particles)
+            for atom in self.atom_subset:
+                self.atom_subset_mask = self.atom_subset_mask.at[atom].set(1)
+
+        key = current_sampler_state.new_PRNG_key
+
+        nr_of_atoms = current_sampler_state.n_particles
+
+        unitless_displacement_sigma = self.displacement_sigma.value_in_unit_system(
+            unit.md_unit_system
+        )
+        import jax.random as jrandom
+
+        scaled_displacement_vector = (
+            jrandom.normal(key, shape=(nr_of_atoms, 3)) * unitless_displacement_sigma
+        )
+        import copy
+
+        proposed_sampler_state = copy.deepcopy(current_sampler_state)
```

Comment:
There are a lot of `copy.deepcopy()` calls required in every `MCMove` because of the way we structure our in-place updates, and not just of arrays, but of objects like our neighbor list. I worry that this is going to cause significant performance problems if we ever need these methods to be `jit`ted to be fast.

Right now, I'm anticipating that most of our compute time will spent be in `LangevinDynamics` integration steps, but it may be useful to think about the patterns we are forcing here and the implications this will eventually have on our ability to optimize performance down the road.

Reply:
#34 Created an issue so we don't forget.